### ISSUE TITLE
Microsoft.Health.Fhir.Anonymizer.R4.AzureDataFactoryPipeline is broken at head

### ISSUE THREAD

**Issue by theGOTOguy** [microsoft/Tools-for-Health-Data-Anonymization#189]
Configure `FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.AzureDataFactoryPipeline/AzureDataFactorySettings.json` and `FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.AzureDataFactoryPipeline/scripts/ArmTemplate/arm_template_parameters.json` as per the instructions and then try, e.g.:

```powershell
PS /home/you> Connect-AzAccount
PS /home/you> Get-AzSubscription
PS /home/you> Select-AzSubscription -SubscriptionId "YOURS"
PS /home/you/code/FHIR-Tools-for-Anonymization/FHIR/src> $SubscriptionId = "YOURS"
PS /home/you/code/FHIR-Tools-for-Anonymization/FHIR/src> $BatchAccountName = "YOURSdeidbatchacct"
PS /home/you/code/FHIR-Tools-for-Anonymization/FHIR/src> $BatchAccountPoolName = "YOURSDeidentifierPool"
PS /home/you/code/FHIR-Tools-for-Anonymization/FHIR/src> $BatchComputeNodeSize = "Standard_D2_v3"
PS /home/you/code/FHIR-Tools-for-Anonymization/FHIR/src> $ResourceGroupName = "YOURS-fhir-deidentifier"
PS /home/you/code/FHIR-Tools-for-Anonymization/FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.AzureDataFactoryPipeline> .\DeployAzureDataFactoryPipeline.ps1 -SubscriptionId $SubscriptionId -BatchAccountName $BatchAccountName -BatchAccountPoolName $BatchAccountPoolName -BatchComputeNodeSize $BatchComputeNodeSize -ResourceGroupName $ResourceGroupName
```

The pipeline will fail with an unhelpful error message:
```
<Error>
<Code>ResourceNotFound</Code>
<Message>The specified resource does not exist. RequestId:REDACTED Time:REDACTED</Message>
</Error>
```

We have a fork that does still work at 57542d13f401a925f9465cee22de3b42b4024135, so it was broken at some point after that.

**Comment by moria97** [comment]
Seems there were no updates in the arm template except the location change from fixed "West US" to current location. Can you double check the referenced resources exist and in the correct location?

**Comment by theGOTOguy** [comment]
The only three files I changed for my deployment were:

```
	modified:   FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.AzureDataFactoryPipeline/AzureDataFactorySettings.json
	modified:   FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json
	modified:   FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.AzureDataFactoryPipeline/scripts/ArmTemplate/arm_template_parameters.json
```

Using the exact same three files at 57542d1 vs. head, I received the error above at head and correct behavior at 5754d21.

My belief is that one of the dependency changes was somehow breaking.

**Comment by moria97** [comment]
> The only three files I changed for my deployment were:
> 
> ```
> 	modified:   FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.AzureDataFactoryPipeline/AzureDataFactorySettings.json
> 	modified:   FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json
> 	modified:   FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.AzureDataFactoryPipeline/scripts/ArmTemplate/arm_template_parameters.json
> ```
> 
> Using the exact same three files at [57542d1](https://github.com/microsoft/Tools-for-Health-Data-Anonymization/commit/57542d13f401a925f9465cee22de3b42b4024135) vs. head, I received the error above at head and correct behavior at [57542d1](https://github.com/microsoft/Tools-for-Health-Data-Anonymization/commit/57542d13f401a925f9465cee22de3b42b4024135).
> 
> My belief is that one of the dependency changes was somehow breaking.

You mean the deploy succeeds with [57542d1](https://github.com/microsoft/Tools-for-Health-Data-Anonymization/commit/57542d13f401a925f9465cee22de3b42b4024135) but fails with HEAD? Can you still succeed to deploy with the old code now?

**Comment by theGOTOguy** [comment]
Yes,  I am able to successfully deploy and use the ADF pipelinie at 57542d1.  I am able to deploy from HEAD using the exact same configuration files, but when I actually run the deployed pipeline I get the error:

```
<Error>
<Code>ResourceNotFound</Code>
<Message>The specified resource does not exist. RequestId:REDACTED Time:REDACTED</Message>
</Error>
```

**Comment by moria97** [comment]
Can you attach the configuration file you changed or describe what you have changed?

**Comment by theGOTOguy** [comment]
FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.AzureDataFactoryPipeline/AzureDataFactorySettings.json
```
{
  "dataFactoryName": "REDACTED",
  "resourceLocation": "EastUS",
  "sourceStorageAccountName": "REDACTED",
  "sourceStorageAccountKey": "REDACTED",
  "destinationStorageAccountName": "REDACTED",
  "destinationStorageAccountKey": "REDACTED",
  "sourceStorageContainerName": "REDACTED",
  "destinationStorageContainerName": "REDACTED",
  "sourceContainerFolderPath": "Identified",
  "destinationContainerFolderPath": "Deidentified",
  "activityContainerName": "REDACTED"
}
```

FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.CommandLineTool/configuration-sample.json:
```
{
  "fhirVersion": "R4",
  "processingErrors": "skip",
  "fhirPathRules": [
    {"path": "nodesByType('Extension')", "method": "keep"},

    {"path": "Organization.identifier", "method": "keep"},
    {"path": "Organization.telecom", "method": "keep"},
    {"path": "Organization.address", "method": "keep"},
    {"path": "OrganizationAffiliation.identifier", "method": "keep"},
    {"path": "OrganizationAffiliation.telecom", "method": "keep"},
    {"path": "nodesByType('HumanName').use", "method": "keep"},
    {"path": "nodesByType('Address').country", "method": "keep"},
    {"path": "nodesByType('Address').state", "method": "keep"},
    {"path": "Resource.id", "method": "cryptoHash"},
    {"path": "Bundle.entry.fullUrl", "method": "redact"},
    {"path": "Bundle.entry.request.url", "method": "redact"},
    {"path": "Bundle.entry.response.location", "method":"redact"},
    {"path": "PractitionerRole.notAvailable.description", "method": "redact"},
    {"path": "PractitionerRole.availabilityExceptions", "method": "redact"},
    {"path": "Group.name", "method": "redact"},
    {"path": "Group.characteristic.value", "method": "redact"},
    {"path": "HealthcareService.name", "method": "redact"},
    {"path": "HealthcareService.comment", "method": "redact"},     
    {"path": "HealthcareService.availabilityExceptions", "method": "redact"},    
    {"path": "HealthcareService.extraDetails", "method": "redact"},   
    {"path": "HealthcareService.eligibility.comment", "method": "redact"},   
    {"path": "HealthcareService.notAvaliable.description", "method": "redact"},
    {"path": "Endpoint.name", "method": "redact"},
    {"path": "Endpoint.address", "method": "redact"},
    {"path": "Location.name", "method": "redact"},
    {"path": "Location.alias", "method": "redact"},
    {"path": "Location.description", "method": "redact"},
    {"path": "Location.position", "method": "redact"},
    {"path": "Location.availabilityExceptions", "method": "redact"},   
    {"path": "Substance.description", "method": "redact"},
    {"path": "Device.udiCarrier", "method": "redact"},
    {"path": "Device.distinctIdentifier", "method": "redact"},
    {"path": "Device.serialNumber", "method": "redact"},
    {"path": "Device.deviceName", "method": "redact"},
    {"path": "Task.description", "method": "redact"},
    {"path": "Task.input.value", "method": "redact"},
    {"path": "Task.output.value", "method": "redact"},
    {"path": "Appointment.description", "method": "redact"},
    {"path": "Appointment.comment", "method": "redact"},
    {"path": "Appointment.patientInstruction", "method": "redact"},
    {"path": "AppointmentResponse.comment", "method": "redact"},
    {"path": "Schedule.comment", "method": "redact"},
    {"path": "Slot.comment", "method": "redact"},
    {"path": "VerificationResult.attestation.sourceIdentityCertificate", "method": "redact"},
    {"path": "VerificationResult.attestation.proxyIdentityCertificate", "method": "redact"},
    {"path": "List.title", "method": "redact"},
    {"path": "Library.publisher", "method": "redact"},
    {"path": "Library.usage", "method": "redact"},
    {"path": "AllergyIntolerance.onset", "method": "redact"},
    {"path": "AllergyIntolerance.reaction.description", "method": "redact"},
    {"path": "AdverseEvent.suspectEntity.causality.productRelatedness", "method": "redact"},
    {"path": "Condition.onset", "method": "redact"},
    {"path": "Condition.abatement", "method": "redact"},
    {"path": "Procedure.performed", "method": "redact"},
    {"path": "FamilyMemberHistory.name", "method": "redact"},
    {"path": "FamilyMemberHistory.born", "method": "redact"},
    {"path": "FamilyMemberHistory.age", "method": "redact"},
    {"path": "FamilyMemberHistory.deceased", "method": "redact"},
    {"path": "FamilyMemberHistory.condition.onset as Age", "method": "perturb", "span": 0.1, "rangeType": "proportional", "roundTo": 0},
    {"path": "FamilyMemberHistory.condition.onset", "method": "redact"},
    {"path": "ClinicalImpression.description", "method": "redact"},  
    {"path": "ClinicalImpression.summary", "method": "redact"},
    {"path": "ClinicalImpression.finding.basis", "method": "redact"},
    {"path": "DetectedIssue.detail", "method": "redact"},
    {"path": "Observation.value", "method": "keep"},
    {"path": "Observation.referenceRange", "method": "redact"},
    {"path": "Observation.component.value", "method": "redact"},
    {"path": "Observation.effectiveDateTime", "method": "dateshift", "dateShiftKey": "", "dateShiftScope": "resource", "dateShiftRange": 5},
    {"path": "Media.deviceName", "method": "redact"},
    {"path": "DiagnosticReport.media.comment", "method": "redact"},
    {"path": "DiagnosticReport.conclusion", "method": "redact"},
    {"path": "Specimen.processing.description", "method": "redact"},
    {"path": "Specimen.collection.collectedDateTime", "method": "dateshift", "dateShiftKey": "", "dateShiftScope": "resource", "dateShiftRange": 5},
    {"path": "Specimen.processing.timeDateTime", "method": "dateshift", "dateShiftKey": "", "dateShiftScope": "resource", "dateShiftRange": 5},
    {"path": "Specimen.processing.receivedTime", "method": "dateshift", "dateShiftKey": "", "dateShiftScope": "resource", "dateShiftRange": 5},
    {"path": "Specimen.container.description", "method": "redact"},
    {"path": "BodyStructure.description", "method": "redact"},
    {"path": "ImagingStudy.description", "method": "redact"},
    {"path": "ImagingStudy.series.description", "method": "redact"},
    {"path": "ImagingStudy.series.instance.title", "method": "redact"},
    {"path": "QuestionnaireResponse.item.text", "method": "redact"},
    {"path": "QuestionnaireResponse.item.answer", "method": "redact"},
    {"path": "QuestionnaireResponse.item.item", "method": "redact"},
    {"path": "MedicationAdministration.dosage.text", "method": "redact"},
    {"path": "Immunization.occurrence", "method": "redact"},
    {"path": "ImmunizationEvaluation.description", "method": "redact"},
    {"path": "ImmunizationRecommendation.recommendation.description", "method": "redact"},
    {"path": "CarePlan.title", "method": "redact"},
    {"path": "CarePlan.description", "method": "redact"},
    {"path": "CarePlan.activity.detail.description", "method": "redact"},
    {"path": "CarePlan.activity.detail.scheduled", "method": "redact"},
    {"path": "CareTeam.name", "method": "redact"},
    {"path": "Goal.statusReason", "method": "redact"},
    {"path": "Goal.target.detail", "method": "redact"},
    {"path": "ServiceRequest.patientInstruction", "method": "redact"},
    {"path": "NutritionOrder.oralDiet.instruction", "method": "redact"},
    {"path": "NutritionOrder.supplement.instruction", "method": "redact"},
    {"path": "NutritionOrder.enteralFormula.administrationInstruction", "method": "redact"},
    {"path": "RiskAssessment.prediction.rationale", "method": "redact"},
    {"path": "RiskAssessment.mitigation", "method": "redact"},
    {"path": "RequestGroup.action.prefix", "method": "redact"},
    {"path": "RequestGroup.action.title", "method": "redact"},
    {"path": "RequestGroup.action.description", "method": "redact"},
    {"path": "RequestGroup.action.textEquivalent", "method": "redact"},
    {"path": "RequestGroup.action.timing", "method": "redact"},
    {"path": "RequestGroup.action.action", "method": "redact"},
    {"path": "Communication.payload", "method": "redact"},
    {"path": "CommunicationRequest.payload", "method": "redact"},
    {"path": "Coverage.subscriberId", "method": "redact"},
    {"path": "Coverage.dependent", "method": "redact"},
    {"path": "Coverage.class.value", "method": "redact"},
    {"path": "Coverage.class.name", "method": "redact"},
    {"path": "Coverage.network", "method": "redact"},
    {"path": "CoverageEligibilityRequest.insurance.businessArrangement", "method": "redact"},
    {"path": "CoverageEligibilityResponse.disposition", "method": "redact"},
    {"path": "CoverageEligibilityResponse.insurance.item.name", "method": "redact"},
    {"path": "CoverageEligibilityResponse.insurance.item.description", "method": "redact"},
    {"path": "CoverageEligibilityResponse.insurance.item.benefit.allowed", "method": "redact"},
    {"path": "CoverageEligibilityResponse.insurance.item.benefit.used", "method": "redact"},
    {"path": "CoverageEligibilityResponse.preAuthRef", "method": "redact"},
    {"path": "EnrollmentResponse.disposition", "method": "redact"},
    {"path": "Claim.supportingInfo.value", "method": "redact"},
    {"path": "Claim.insurance.businessArrangement", "method": "redact"},
    {"path": "Claim.insurance.preAuthRef", "method": "redact"},
    {"path": "ClaimResponse.disposition", "method": "redact"},
    {"path": "ClaimResponse.preAuthRef", "method": "redact"},
    {"path": "ClaimResponse.processNote.text", "method": "redact"},
    {"path": "ClaimResponse.insurance.businessArrangement", "method": "redact"},
    {"path": "Invoice.cancelledReason", "method": "redact"},
    {"path": "PaymentReconciliation.disposition", "method": "redact"},
    {"path": "PaymentReconciliation.processNote.text", "method": "redact"},
    {"path": "Account.name", "method": "redact"},
    {"path": "Account.description", "method": "redact"},
    {"path": "ChargeItem.overrideReason", "method": "redact"},
    {"path": "ChargeItemDefinition.title", "method": "redact"},
    {"path": "ChargeItemDefinition.publisher", "method": "redact"},
    {"path": "ChargeItemDefinition.applicability", "method": "redact"},
    {"path": "ChargeItemDefinition.propertyGroup.applicability", "method": "redact"},
    {"path": "Contract.name", "method": "redact"},
    {"path": "Contract.title", "method": "redact"},
    {"path": "Contract.subtitle", "method": "redact"},
    {"path": "Contract.alias", "method": "redact"},
    {"path": "Contract.term.text", "method": "redact"},
    {"path": "Contract.term.offer.answer.value", "method": "redact"},
    {"path": "Contract.term.offer.text", "method": "redact"},
    {"path": "Contract.term.offer.linkId", "method": "redact"},
    {"path": "Contract.term.asset.context.text", "method": "redact"},
    {"path": "Contract.term.asset.condition", "method": "redact"},
    {"path": "Contract.term.asset.text", "method": "redact"},
    {"path": "Contract.term.asset.linkId", "method": "redact"},
    {"path": "Contract.term.asset.answer", "method": "redact"},
    {"path": "Contract.term.asset.valuedItem.payment", "method": "redact"},
    {"path": "Contract.term.asset.valuedItem.linkId", "method": "redact"},
    {"path": "Contract.term.action.linkId", "method": "redact"},
    {"path": "Contract.term.action.contextLinkId", "method": "redact"},
    {"path": "Contract.term.action.requesterLinkId", "method": "redact"},
    {"path": "Contract.term.action.performerLinkId", "method": "redact"},
    {"path": "Contract.term.action.reason", "method": "redact"},
    {"path": "Contract.term.action.reasonLinkId", "method": "redact"},
    {"path": "Contract.term.group", "method": "redact"},
    {"path": "ExplanationOfBenefit.disposition", "method": "redact"},
    {"path": "ExplanationOfBenefit.preAuthRef", "method": "redact"},
    {"path": "ExplanationOfBenefit.supportingInfo.value", "method": "redact"},
    {"path": "ExplanationOfBenefit.insurance.preAuthRef", "method": "redact"},
    {"path": "ExplanationOfBenefit.accident.location", "method": "redact"},
    {"path": "ExplanationOfBenefit.item.location", "method": "redact"},
    {"path": "ExplanationOfBenefit.addItem.location", "method": "redact"},
    {"path": "ExplanationOfBenefit.processNote.text", "method": "redact"},
    {"path": "ExplanationOfBenefit.benefitBalance.name", "method": "redact"},
    {"path": "ExplanationOfBenefit.benefitBalance.description", "method": "redact"},
    {"path": "ExplanationOfBenefit.benefitBalance.financial.allowed", "method": "redact"},
    {"path": "InsurancePlan.name", "method": "redact"},
    {"path": "InsurancePlan.alias", "method": "redact"},
    {"path": "InsurancePlan.coverage.benefit.requirement", "method": "redact"},
    {"path": "InsurancePlan.plan.generalCost.comment", "method": "redact"},
    

    {"path": "nodesByType('Expression').description", "method": "redact"},   
    {"path": "nodesByType('Expression').expression", "method": "redact"},    
    {"path": "nodesByType('Attachment').title", "method": "redact"},
    {"path": "nodesByType('Attachment').url", "method": "redact"},
    {"path": "nodesByType('CodeableConcept').text", "method": "keep"},
    {"path": "nodesByType('Coding').display", "method": "keep"},
    {"path": "nodesByType('Coding').code", "method": "keep"},
    {"path": "nodesByType('ContactDetail').name", "method": "redact"},
    {"path": "nodesByType('Identifier').value", "method": "cryptoHash"},
    {"path": "nodesByType('Reference').display", "method": "cryptoHash"},
    {"path": "nodesByType('Reference').reference", "method": "cryptoHash"},

    {"path": "nodesByType('Narrative')", "method": "redact"},
    {"path": "nodesByType('base64Binary')", "method": "redact"},
    {"path": "nodesByType('date')", "method": "dateshift"},
    {"path": "nodesByType('dateTime')", "method": "dateshift"},
    {"path": "nodesByType('instant')", "method": "dateshift"},
    {"path": "nodesByType('Address')", "method": "redact"},
    {"path": "nodesByType('Age')", "method": "redact"},
    {"path": "nodesByType('Annotation')", "method": "redact"},
    {"path": "nodesByType('HumanName')", "method": "redact"},
    {"path": "nodesByType('ContactPoint')", "method": "redact"}
  ],
  "parameters": {
    "dateShiftKey": "",
    "dateShiftScope": "resource",
    "cryptoHashKey": "REDACTED",
    "encryptKey": "",
    "enablePartialAgesForRedact": true,
    "enablePartialDatesForRedact": true,
    "enablePartialZipCodesForRedact": true,
    "restrictedZipCodeTabulationAreas": [
      "036",
      "059",
      "102",
      "203",
      "205",
      "369",
      "556",
      "692",
      "821",
      "823",
      "878",
      "879",
      "884",
      "893"
    ]
  }
}
```

FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.AzureDataFactoryPipeline/scripts/ArmTemplate/arm_template_parameters.json:
```
{
  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "location": {
      "value": "EastUS"
    },
    "factoryName": {
      "value": "fhirdeidentifier"
    },
    "sourceStorageLinkedService_connectionString": {
      "value": "REDACTED"
    },
    "sourceStorageActivityApplicationContainer": {
      "value": "fhirdeidentifiercontainer"
    },
    "sourceStorageContainerName": {
      "value": "REDACTED"
    },
    "sourceContainerFolderPath": {
      "value": "Identified/"
    },
    "destinationStorageLinkedService_connectionString": {
      "value": "REDACTED"
    },
    "destinationStorageContainerName": {
      "value": "REDACTED"
    },
    "destinationContainerFolderPath": {
      "value": "Deidentified/"
    },
    "azureBatchLinkedService_accessKey": {
      "value": "REDACTED"
    },
    "azureBatchLinkedService_poolName": {
      "value": "REDACTED"
    },
    "azureBatchLinkedService_properties_typeProperties_accountName": {
      "value": "REDACTED"
    },
    "azureBatchLinkedService_properties_typeProperties_batchUri": {
      "value": "REDACTED"
    },
    "fhirVersion": {
      "value": "R4"
    }
  }
}

```

**Comment by moria97** [comment]
> ResourceNotFound

The configs looks good. And we don't see changes happen after the commit [57542d1](https://github.com/microsoft/Tools-for-Health-Data-Anonymization/commit/57542d13f401a925f9465cee22de3b42b4024135). Have you changed the pipeline/arm template in your own HEAD or it's just the latest master branch?

Can you provide more information from this error? Like the pipeline script output?

**Comment by theGOTOguy** [comment]
The procedure was pretty simple.

I needed to re-deploy the pipeline so I took the three files above, copied them into a fresh clone of this repository at HEAD, and deployed as per the instructions.  The pipeline deployed normally, but the run it initiated failed with:

```
<Error>
<Code>ResourceNotFound</Code>
<Message>The specified resource does not exist. RequestId:REDACTED Time:REDACTED</Message>
</Error>
```

To try to diagnose the problem, I took my last known-good fork at 57542d1, copied the exact same three files into the fork, and tried to deploy the pipeline again.

This time the pipeline deployed normally and the run completed as expected.  This leads me to believe that one of the dependency updates was a breaking change and has subtly broken the ADF pipeline.

Question:  Are you able to successfully deploy and run the ADF pipeline on some sample data (say, from [Synthea](https://github.com/synthetichealth/synthea)) at head?  That would probably be the clearest way to for you to confirm this issue.
